extensions/amp-ad-network-fake-impl/0.1/amp-ad-network-fake-impl.js

```diff
@@ -106,55 +112,191 @@ export class AmpAdNetworkFakeImpl extends AmpA4A {
+            }
+            lastRuntimeElement = child;
+            if (child.hasAttribute('custom-element')) {
+              extensions.push({
+                'custom-element': child.getAttribute('custom-element'),
+                'src': child.getAttribute('src'),
+              });
+            }
+            if (child.hasAttribute('custom-template')) {
+              extensions.push({
+                'custom-template': child.getAttribute('custom-template'),
+                'src': child.getAttribute('src'),
+              });
+            }
+          }
+        }
+        if (
+          child.tagName == 'LINK' &&
+          child.getAttribute('rel') == 'stylesheet' &&
+          child.hasAttribute('href')
+        ) {
+          const styleJson = {href: child.getAttribute('href')};
+          if (child.hasAttribute('media')) {
+            styleJson.media = child.getAttribute('media');
+          }
+          styles.push(styleJson);
+        }
+        if (this.element.hasAttribute('amp-story')) {
+          if (
+            child.tagName == 'META' &&
+            child.hasAttribute('name') &&
+            child.hasAttribute('content')
+          ) {
+            if (child.getAttribute('name') == 'amp-cta-type') {
+              ctaType = child.getAttribute('content');
+            }
+            if (child.getAttribute('name') == 'amp-cta-url') {
+              ctaUrl = child.getAttribute('content');
+            }
+          }
+        }
       }
-      doc.head.removeChild(script);
     }
-
-    // Remove boilerplate styles.
-    const styles = doc.head.querySelectorAll('style[amp-boilerplate]');
-    for (let i = 0; i < styles.length; i++) {
-      const style = styles[i];
-      style.parentNode.removeChild(style);
+    const imgMetadata = [];
+    const imgs = doc.querySelectorAll('amp-img[src]');
+    for (let i = 0; i < imgs.length; i++) {
+      const img = imgs[i];
+      let width,
+        height,
+        area = -1;
+      if (img.hasAttribute('width')) {
+        width = img.getAttribute('width');
+      }
+      if (img.hasAttribute('height')) {
+        height = img.getAttribute('height');
+      }
+      if (height && width) {
+        area = height * width;
+      }
+      imgMetadata.push({
+        src: img.getAttribute('src'),
+        area,
+      });
     }
 
-    let creative = root./*OK*/ outerHTML;
-
-    // Metadata
-    creative += '<script type="application/json" amp-ad-metadata>';
-    creative += '{';
-
-    if (this.element.hasAttribute('amp-story')) {
-      const ctaTypeMeta = root.querySelector('[name=amp-cta-type]');
-      const ctaType = ctaTypeMeta && ctaTypeMeta.content;
-      creative += `"ctaType": "${ctaType}", `;
+    const json = doc.querySelectorAll('script[type]');
+    for (let i = 0; i < json.length; i++) {
+      const script = json[i];
+      const type = script.getAttribute('type');
+      if (
+        type == 'application/json' &&
+        script.hasAttribute('id') &&
```

Comment:
What happens if `amp-ad-metadata` element has an id?

Reply:
It shouldn't, since the `amp-ad-metadata` script is always generated the same way, but I added a condition just in case.